cc_macro.c

```diff
@@ -424,6 +485,44 @@ int macro_expression(void)
 	return macro_bitwise_expr();
 }
 
+void handle_function_like_macro(struct macro_list* hold)
+{
+
+	eat_current_token(); /* Skip '(' */
+	struct macro_argument* argument = calloc(1, sizeof(struct macro_argument));
+	hold->arguments = argument;
+
+	argument->next = NULL;
+	/* A NULL name in the first argument means a function macro without any arguments */
+	argument->name = NULL;
+
+	while(macro_token->s[0] != ')')
+	{
+		/* ... is not tokenized as a single token here */
+		if(macro_token->s[0] == '.' &&
+			macro_token->next->s[0] == '.' &&
+			macro_token->next->next->s[0] == '.')
+		{
+			line_error_token(macro_token);
+			fputs("Variadic function-like macros not supported.", stderr);
+			exit(EXIT_FAILURE);
+		}
+
+		argument->name = macro_token->s;
+		eat_current_token(); /* skip past name to comma */
+
+		if(macro_token->s[0] == ',')
+		{
+			argument->next = calloc(1, sizeof(struct macro_argument));
+			argument = argument->next;
```

Comment:
possibly here too

Reply:
None of the other uses of `malloc`/`calloc` checked for `NULL`. If we want to exhaustively check it might be better to create an `xcalloc`/`xmalloc` function that checks internally and exclusively use that?

In general I fail to see what can really be done about OOM other than exiting.

Reply:
Oh yes, let's not worry about OOM. I'm mostly looking at not crashing M2-Planet on illegal C input files.